bmm and contiguous constraints

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81714
* __->__ #81527

- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

migeed-z committed:
bmm and contiguous constraints

[ghstack-poisoned]

migeed-z committed:
Update on "bmm and contiguous constraints"


- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

[ghstack-poisoned]

migeed-z committed:
Update on "bmm and contiguous constraints"


- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

[ghstack-poisoned]

migeed-z committed:
Update on "bmm and contiguous constraints"


- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

[ghstack-poisoned]

jamesr66a commented on torch/fx/experimental/migrate_gradual_types/constraint_generator.py:
```diff
@@ -57,6 +57,48 @@ def get_attr_inference_rule(n: Node, symbols, constraints, counter):
+    inputs_are_tensors = Conj([BinConstraintT(bmm_input1, TensorType(dims_input1), op_eq),
+                               BinConstraintT(bmm_input2, TensorType(dims_input2), op_eq),
+                               BinConstraintT(bmm_output, TensorType([batch_size, dims_input1[1], dims_input2[2]]), op_eq),
+                               *consistency_constraints, DGreatestUpperBound(batch_size, dims_input1[0], dims_input2[0])])
```
Operator overloads would really clean this up :)

migeed-z commented:
@pytorchbot merge -g

migeed-z committed:
Update on "bmm and contiguous constraints"


- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g

migeed-z committed:
Update on "bmm and contiguous constraints"


- Batch matrix multiplication constraints which generate constraints for dimensions according to the rules of batch matrix multiplication 
- Contiguous constraints which are just input = output

[ghstack-poisoned]

migeed-z commented:
@pytorchbot merge -g
